pkg/plugins/policies/meshratelimit/plugin/v1alpha1/plugin_test.go

```diff
@@ -0,0 +1,458 @@
+									OnRateLimit: &api.OnRateLimit{
+										Status: policies_xds.PointerOf(uint32(444)),
+										Headers: []api.HeaderValue{
+											{
+												Key:    "x-kuma-rate-limit-header",
+												Value:  "test-value",
+												Append: policies_xds.PointerOf(true),
+											},
+											{
+												Key:   "x-kuma-rate-limit",
+												Value: "other-value",
+											},
+										},
+									},
+								},
+							},
+						},
+					}},
+				},
+			}
+
+			resources := xds_context.NewResources()
+			resources.MeshLocalResources[core_mesh.MeshGatewayType] = GatewayResources()
+			resources.MeshLocalResources[core_mesh.MeshGatewayRouteType] = GatewayRoutes()
+
+			context := createSimpleMeshContextWith(resources)
+			proxy := core_xds.Proxy{
+				APIVersion: "v3",
+				Dataplane: &core_mesh.DataplaneResource{
+					Meta: &test_model.ResourceMeta{
+						Mesh: "default",
+						Name: "gateway",
+					},
+					Spec: &mesh_proto.Dataplane{
+						Networking: &mesh_proto.Dataplane_Networking{
+							Address: "127.0.0.1",
+							Gateway: &mesh_proto.Dataplane_Networking_Gateway{
+								Tags: map[string]string{
+									mesh_proto.ServiceTag: "gateway",
+								},
+								Type: mesh_proto.Dataplane_Networking_Gateway_BUILTIN,
+							},
+						},
+					},
+				},
+				Policies: core_xds.MatchedPolicies{
+					Dynamic: map[core_model.ResourceType]core_xds.TypedMatchingPolicies{
+						api.MeshRateLimitType: {
+							Type:      api.MeshRateLimitType,
+							FromRules: fromRules,
+						},
+					},
+				},
+			}
+			gatewayGenerator := gatewayGenerator()
+			generatedResources, err := gatewayGenerator.Generate(context, &proxy)
+			if err != nil {
+				return
+			}
+			plugin := plugin.NewPlugin().(core_plugins.PolicyPlugin)
+
+			Expect(plugin.Apply(generatedResources, context, &proxy)).To(Succeed())
+			Expect(util_proto.ToYAML(generatedResources.ListOf(envoy_resource.RouteType)[0].Resource)).To(test_matchers.MatchGoldenYAML(filepath.Join("testdata", "gateway_basic_routes.golden.yaml")))
+			Expect(util_proto.ToYAML(generatedResources.ListOf(envoy_resource.ListenerType)[0].Resource)).To(test_matchers.MatchGoldenYAML(filepath.Join("testdata", "gateway_basic_listener.golden.yaml")))
+		})
+})
+
+func GatewayResources() *core_mesh.MeshGatewayResourceList {
+	return &core_mesh.MeshGatewayResourceList{
+		Items: []*core_mesh.MeshGatewayResource{{
+			Meta: &test_model.ResourceMeta{Name: "gateway", Mesh: "default"},
+			Spec: &mesh_proto.MeshGateway{
+				Selectors: []*mesh_proto.Selector{{
+					Match: map[string]string{
+						mesh_proto.ServiceTag: "gateway",
+					}},
+				},
+				Conf: &mesh_proto.MeshGateway_Conf{
+					Listeners: []*mesh_proto.MeshGateway_Listener{
+						{
+							Protocol: mesh_proto.MeshGateway_Listener_HTTP,
+							Port:     8080,
+						},
+					},
+				},
+			},
+		}},
+	}
+}
+
+func GatewayRoutes() *core_mesh.MeshGatewayRouteResourceList {
```

Comment:
maybe you could reuse [samples](https://github.com/kumahq/kuma/blob/52c992533bb8ab5f565e56f1b7a9af0a96e361a7/pkg/test/resources/samples/gateway_samples.go)